2-Draft-Accepted/RFCxxxx-Secrets-Management.md

```diff
@@ -0,0 +1,150 @@
+New-PSSession -ComputerName myServer -Credential (Get-Secret -Name MyCreds)
+```
+
+## Specification
+
+### Secrets Vault
+
+There are two types of vaults: local and remote.
+A local vault, by default, is already created and named `Default`.
+`Default` on Windows is [CredMan](https://docs.microsoft.com/en-us/windows/desktop/SecAuthN/credentials-management).
+`Default` on macOS is [KeyChain](https://developer.apple.com/documentation/security/keychain_services).
+
+>[!NOTE]
+>KeyChain support is unlikely to be available in the first release of this feature.
+
+`Default` on Linux is [Hashicorp Vault](https://www.vaultproject.io/).
+
+>[!NOTE]
+>Linux has many options for local credential management, additional support may be
+>added in the future or via community provided extensions.
+
+### Credential Vault Extensions
+
+Vault extensions whether they are local or remote are provided as modules.
+
+Extensions are modules that provide a `Get-Secret` cmdlet that accepts a
+`-Credential` or `-AccessToken` parameter for authentication.
+Expectation is that each module will have their own cmdlet and not literally
+named `Get-Secret`.
+
+Extensions register themselves by calling `Register-SecretsVaultExtension`
+passing the name of the relevant cmdlet to `-Cmdlet`, the module name to `-Module`,
+and the module version to `-ModuleVersion`.
+This can be done as part of importing the module or explicitly by the user.
+
+`Register-SecretsVaultExtension` has a `-Default` switch to register the
+default local vault (if applicable, see above) if it has been unregistered.
+
+`Register-SecretsVaultExtension` takes a `-AccessToken` or `-Credential` parameter
+for authenticating with the remote vault.
+If neither are supplied, then the current user context is used and relies on
+the extension cmdlet to handle authentication.
+
+`Unregister-SecretsVaultExtension` should be used to remove any existing
+registration or when updating a registration to a new version.
+
+`Get-SecretsVaultExtension` will enumerate registered extensions returning
+the module name, module version, and cmdlet used.
+
+### Storing Secrets
+
+The `Register-Secret` cmdlet is used to store a secret.
+The `-Name` must be unique across all secrets regardless if local or remote.
+A `-Secret` parameter accepts an object that can be either a PSCredential
+or SecureString.
+
+### Retrieving Secrets
+
+The `Get-Secret` cmdlet is used to retrieve secrets.
+The `-Name` parameter is mandatory and retrieves the secret associated with
+that name.
+The returned object will either be a PSCredential or SecureString.
+
+### Authorization
+
+Access to the credential vault is always using the current process security context.
+In the case of remote vaults, the remote credential is stored within the local
+default vault and retrieved as needed when accessing secrets from that remote
+vault.
+
+### Secrets Supported
+
+Secrets supported will be:
+
+- PSCredential
+- SecureString
```

Comment:
The .Net Core team advises to not use SecureString any more for new development, mainly because it is not secure and the implementation is a no-op on *nix
https://docs.microsoft.com/en-us/dotnet/api/system.security.securestring

Reply:
@bergmeister i found this comment related. 

> Early on, the @PowerShell/powershell-committee discussed introducing a `SensitiveString` to replace the functional need of `SecureString` even though both are not secure (the `SecureString` type would still be needed for backwards compat). A type (whether "Sensitive" or "Secure" is needed to indicate to PowerShell to prompt without echoing the input so it's used more than just for remoting. As for the original issue of this bug, this has been fixed (you don't get an error anymore), just keep in mind the SecureString is internally in plain text.

Source : https://github.com/PowerShell/PowerShell/issues/1654#issuecomment-474987708

Reply:
Additionally, PSCredential currently still uses SecureString, so unless we want to venture a breaking change on that API we'd need to support it regardless.

Reply:
Yes, it's being used here for compatibility reasons.  But even without encryption support on non-Windows, you get the benefit of having the string cleared out of memory.

Reply:
-AsPlaintext should also be an option, to avoid unnecessary conversions from SecureString when you're dealing with things like API Keys or Connection Strings.

Reply:
I've had cmdlet's for a very long time that fill this purpose (~psv2).  They support four data types:
[PSCredential], [SecureString], [string], and [Hashtable]

I believe this should be supported here as well, as they've all proved valuable.  [PSCredential] and [SecureString] are required by many commands.  [string] works for access tokens and the like, which are rarely accepted via cmdlets as [SecureString].  [Hashtable] is stored as a [string], but incredibly useful when one wants to secure an entire thing to splat and just rehydrate it.

Reply:
Will add support for [string], [hashtable], and [byte[]].